### PR TITLE
Add MangaPark, Mangago, Tapastic, nhentai, E-Hentai and Sen Manga sources

### LINKED ISSUE #1: mangapark
can you add mangapark.me as a source.

### LINKED ISSUE #6: Mangago
Can you please add Mangago.me

http://www.mangago.me

### LINKED ISSUE #28: raw.senmanga.com
//Would soneone please add a raws source such as
https://raw.senmanga.com
// as was asked here :
//https://github.com/inorichi/tachiyomi/issues/404

### LINKED ISSUE #30: Tapastick
https://tapastic.com/browse?sortType=SUBSCRIBE&browse=ALL

### PR DESCRIPTION
**Pull request by null-dev**
This PR is mostly done and ready for review! It adds some sources that have been requested by users.

**Changes in this PR:**
Add MangaPark source.
Fixes #1.

Add Mangago source.
Fixes #6.

Add Tapastic source.
Fixes #30.

Add nhentai source.
Please note that this source consists of multiple source classes in one apk. Tachiyomi will not be able to load this source until https://github.com/inorichi/tachiyomi/pull/761 is merged.

Add E-Hentai source.
Please note that this source consists of multiple source classes in one apk. Tachiyomi will not be able to load this source until https://github.com/inorichi/tachiyomi/pull/761 is merged.

Add Sen Manga source.
Fixes #28 and inorichi/tachiyomi#404.

### CODE REVIEW

File: src/en/tapastic/src/eu/kanade/tachiyomi/extension/en/tapastic/Tapastic.kt [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        //Chapters are stored in JavaScript as JSON!
+        = response.asJsoup().getElementsByTag("script").filter {
+            it.data().trim().startsWith("var _data")
+        }.flatMap {
+            val text = it.data()
+            val episodeVar = text.indexOf("episodeList")
+            if(episodeVar == -1)
+                return@flatMap emptyList<SChapter>()
+
+            val episodeLeftBracket = text.indexOf('[', startIndex = episodeVar)
```
Comment (null-dev) [inline src/en/tapastic/src/eu/kanade/tachiyomi/extension/en/tapastic/Tapastic.kt]: TODO: Deal with titles that have square brackets in them.

**Comment by j2ghz** [comment]
About tapastick, I think it would be more user friendly if the search cancelled the filters, not the other way around as it is now, if it's possible.

**Comment by null-dev** [comment]
Mangago also does not support search + filter at the same time. Should I change its behavior as well?

**Comment by j2ghz** [comment]
I think so, it can be applied to any source. The only one I tested was tapastic.

File: build.gradle [hunk 2]
```diff
@@ -1,5 +1,5 @@
 buildscript {
-    ext.kotlin_version = '1.0.6'
+    ext.kotlin_version = '1.1.1'
```
Comment (null-dev) [inline build.gradle]: I have verified that all the sources successfully compile on Kotlin 1.1.1.

**Comment by null-dev** [comment]
Please note that I am aware of the download percentage bug in the Sen Manga source but I am unable to devise a fix for it currently as the server does not appear to send a `Content-Length` header.

**Comment by rouji4256** [comment]
Thanks man! Being able to read and download from Sen Manga is already better than nothing, thanks again!

**Comment by j2ghz** [comment]
Is this ready?

**Comment by null-dev** [comment]
I think it is.

The reason there are so many code cleanup commits is because I always seem to notice ugly code a week after I wrote it and so I go back and fix it. Then a week later I notice more ugly code and I go fix it again and the cycle continues :cry:.

**Pull request merged**